src/transport/PeerMessageCounter.h

```diff
@@ -85,7 +86,65 @@ class PeerMessageCounter
         return CHIP_NO_ERROR;
     }
 
-    CHIP_ERROR Verify(uint32_t counter) const
+    /**
+     * @brief Implementation of spec 4.5.4.2
+     *
+     * For encrypted messages of Group Session Type, any arriving message with a counter in the range
+     * [(max_message_counter + 1) to (max_message_counter + 2^31 - 1)] (modulo 2^32) SHALL be considered
+     * new, and cause the max_message_counter value to be updated. Messages with counters from
+     * [(max_message_counter - 2^31) to (max_message_counter - MSG_COUNTER_WINDOW_SIZE - 1)] (modulo 2^
+     * 32) SHALL be considered duplicate. Message counters within the range of the bitmap SHALL be
+     * considered duplicate if the corresponding bit offset is set to true.
+     *
+     */
+    CHIP_ERROR VerifyGroupCounter(uint32_t counter)
+    {
+        uint32_t highLimit = mSynced.mMaxCounter + kGroupOffset + 1;
+        // 1. Check counter value if in valid range
+        if (highLimit < mSynced.mMaxCounter)
+        {
+            // Rollover possible
+            if ((counter > mSynced.mMaxCounter) || (counter <= highLimit))
+            {
+                return CHIP_NO_ERROR;
+            }
+        }
+        else
+        {
+            if ((counter > mSynced.mMaxCounter) && (counter <= highLimit))
+            {
+                return CHIP_NO_ERROR;
+            }
+        }
+        // 2.  Check counter if in invalid range
+        uint32_t invalidLowerLimit = mSynced.mMaxCounter - kGroupOffset;
+        highLimit                  = mSynced.mMaxCounter - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE - 1;
+        if (highLimit > invalidLowerLimit)
+        {
+            if ((counter >= invalidLowerLimit) && (counter <= highLimit))
+            {
+                return CHIP_ERROR_MESSAGE_COUNTER_OUT_OF_WINDOW;
+            }
+        }
+
+        // 3. Counter Window check
+        uint32_t offset = mSynced.mMaxCounter - counter;
+        if (offset < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
+        {
+            if (mSynced.mWindow.test(offset))
```

Comment:
This should actually be something like:
```suggestion
        if (offset <= CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE)
        {
            if ((offset == 0) || mSynced.mWindow.test(offset-1))
```
as far as I can tell.  Yes, unicast counters get this wrong.  That's tracked in https://github.com/project-chip/connectedhomeip/issues/9969

At least this is what it should be if my understanding is correct that `mMaxCounter` is the "largest" counter value we actually observed.

That said, if you want to leave this wrong in the same way as unicast counters and we get both fixed in a followup that would be OK too.  Might be less churn for this change to do that.

Reply:
Applied